_posts/2016-09-09-rails-multitenant-site.md

```diff
@@ -0,0 +1,167 @@
+
+## Apartment Gem
+
+Ryan Bates has a [great railscast](http://railscasts.com/episodes/389-multitenancy-with-postgresql) on how to create a multi-tenant site using PostgreSQL schemas and search paths. It's surprisingly quite straight forward.
+
+We chose to use the [apartment gem](https://github.com/influitive/apartment) which does a great job of wrapping up and abstracting the logic. It looks after mapping the request coming in to the correct database schema or 'tenant' as it calls them. It also ensures that all database migrations are run or rolled back on all your schemas.
+
+### Setup
+
+Rather than going into detail on the setup instructions, there's an excellent screencast from Chris Oliver at [Go Rails](https://gorails.com) on [how setup a multi-tenant site with apartment](https://gorails.com/episodes/multitenancy-with-apartment). 
+
+### Elevators
+
+The apartment gem has the concept of 'Elevators' to switch between tenants. Elevator can be used to to switch the tenant based on the request coming in. Out of the box, requests can be matched by subdomain or host, but you can easily build your own Elevator to switch based on some custom attributes like headers or query params.
+
+Since we have different sites on different domains, we match on hostname, using the `HostHash` elevator. However, by default, this elevator throws an exception if it doesn't find a match. For example, if a user requests `https://foo.somedomain.com` instead of `http://www.somedomain.com`. To handle this a bit better we created a custom Elevator that renders a static 404 page:
+
+```ruby
+class RescuedHostHash < ::Apartment::Elevators::HostHash
+
+  def call(env)
+    super
+  rescue ::Apartment::TenantNotFound => e
+    Rollbar.error(e, "ERROR: Apartment Tenant not found: #{Apartment::Tenant.current.inspect}")
+    static_404_page = File.read(Rails.root.to_s + '/public/404.html')
+    return [ 404,  {'Content-Type' => 'text/html' }, [static_404_page] ]
+  end
+
+end
+```
+
+### Site specific configuration
+
+While both sites were very similar, we did need to be able to customize them both. To do this we stored some site-specific variables in a YAML file, wrapped those values in a simple and made it accessible to via a `current_site` helper method.
+
+```yaml
+# config/site-settings.yml
+somedomain:
+  name:     'Some Domain'
+  email:    'info@somedomain.de'
+  locale:   'de'
+  timezone: 'Europe/Berlin'
+otherdomain:
+  name:     'Other Doman'
+  email:    'info@otherdomain.co.uk'
+  locale:   'en'
+  timezone: 'Europe/London'
+```
+
+We wrap the site settings with a simple ruby class set it using an application controller concern.
```

Comment:
Should it be mentioned that this "simple ruby class" is named `Site`. Would make it easier to understand the following snippet.